File: app/src/main/java/me/blog/korn123/easyphotomap/activities/BatchPopupActivity.java
package me.blog.korn123.easyphotomap.activities;

import android.app.Activity;
import android.content.Context;
import android.graphics.Typeface;
import android.location.Address;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.beardedhen.androidbootstrap.TypefaceProvider;
import com.drew.imaging.jpeg.JpegMetadataReader;
import com.drew.metadata.Metadata;
import com.drew.metadata.exif.ExifSubIFDDirectory;
import com.drew.metadata.exif.GpsDirectory;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang.time.StopWatch;

import java.io.File;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;

import butterknife.ButterKnife;
import butterknife.OnClick;
import me.blog.korn123.easyphotomap.R;
import me.blog.korn123.easyphotomap.constants.Constant;
import me.blog.korn123.easyphotomap.helper.PhotoMapDbHelper;
import me.blog.korn123.easyphotomap.models.PhotoMapItem;
import me.blog.korn123.easyphotomap.utils.BitmapUtils;
import me.blog.korn123.easyphotomap.utils.CommonUtils;
import me.blog.korn123.easyphotomap.utils.DialogUtils;

/**
 * Created by CHO HANJOONG on 2016-09-11.
 */
public class BatchPopupActivity extends Activity {

    ProgressBar mProgressBar;
    private boolean mEnableUpdate = true;

    int mTotalPhoto = 0;
    int mSuccessCount = 0;
    int mFailCount = 0;
    int mNoGPSInfoCount = 0;
    int mReduplicationCount = 0;
    private int mProgressStatus = 0;

    TextView mInfoText;
    TextView mInfoText2;
    TextView mInfoText3;
    TextView mInfoText4;
    TextView mInfoText5;
    TextView mInfoText6;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TypefaceProvider.registerDefaultIconSets();
        setContentView(R.layout.activity_batch_popup);
        ButterKnife.bind(this);
        mProgressBar = (ProgressBar)findViewById(R.id.progressBar);
        mInfoText = (TextView)findViewById(R.id.infoText);
        mInfoText2 = (TextView)findViewById(R.id.infoText2);
        mInfoText3 = (TextView)findViewById(R.id.infoText3);
        mInfoText4 = (TextView)findViewById(R.id.infoText4);
        mInfoText5 = (TextView)findViewById(R.id.infoText5);
        mInfoText6 = (TextView)findViewById(R.id.infoText6);
        mInfoText.setTypeface(Typeface.DEFAULT);
        mInfoText2.setTypeface(Typeface.DEFAULT);
        mInfoText3.setTypeface(Typeface.DEFAULT);
        mInfoText4.setTypeface(Typeface.DEFAULT);
        mInfoText5.setTypeface(Typeface.DEFAULT);
        mInfoText6.setTypeface(Typeface.DEFAULT);

        ArrayList<String> listImagePath = getIntent().getStringArrayListExtra("listImagePath");
        mTotalPhoto = listImagePath.size();
        Thread thread = new RegisterThread(BatchPopupActivity.this, listImagePath);
        mProgressBar.setMax(mTotalPhoto);
        thread.start();
    }

    @OnClick({R.id.stop, R.id.close})
    public void buttonClick(View view) {
        switch (view.getId()) {
            case R.id.stop:
                mEnableUpdate = false;
                break;
            case R.id.close:
                finish();
                break;
        }
    }

    Handler progressHandler = new Handler(new Handler.Callback() {
        public boolean handleMessage(Message msg) {
            mInfoText.setText(++mProgressStatus + " / " + mTotalPhoto);
            mInfoText2.setText(getString(R.string.batch_popup_message2) + ": " + mSuccessCount);
            mInfoText3.setText(getString(R.string.batch_popup_message3) + ": " + mReduplicationCount);
            mInfoText4.setText(getString(R.string.batch_popup_message4) + ": " + mNoGPSInfoCount);
            mInfoText5.setText(getString(R.string.batch_popup_message5) + ": " + mFailCount);
            mProgressBar.setProgress(mProgressStatus);
            return true;
        }
    });

    class RegisterThread extends Thread {
        ArrayList<String> listImagePath;
        Context context;

        RegisterThread(Context context, ArrayList<String> listImagePath) {
            this.listImagePath = listImagePath;
            this.context = context;
        }

        public void run() {
            for (String imagePath : listImagePath) {
                StopWatch stopWatch = new StopWatch();
                stopWatch.start();
                if (!mEnableUpdate) return;
                Message message = progressHandler.obtainMessage();
                try {
                    String fileName = FilenameUtils.getName(imagePath) + ".origin";
                    File targetFile = null;
                    if (CommonUtils.loadBooleanPreference(BatchPopupActivity.this, "enable_create_copy")) {
                        targetFile = new File(Constant.WORKING_DIRECTORY + fileName);
                        if (!targetFile.exists()) {
                            FileUtils.copyFile(new File(imagePath), targetFile);
                        }
                    } else {
                        targetFile = new File(imagePath);
                        // remove .origin extension
                        fileName = FilenameUtils.getBaseName(fileName);
                    }

                    Metadata metadata = JpegMetadataReader.readMetadata(targetFile);
                    PhotoMapItem item = new PhotoMapItem();
                    item.imagePath = targetFile.getAbsolutePath();
                    ExifSubIFDDirectory exifSubIFDDirectory = metadata.getFirstDirectoryOfType(ExifSubIFDDirectory.class);
                    Date date = exifSubIFDDirectory.getDate(ExifSubIFDDirectory.TAG_DATETIME_ORIGINAL, TimeZone.getDefault());
                    if (date != null) {
                        item.date = CommonUtils.DATE_TIME_PATTERN.format(date);
                    } else {
                        item.date = getString(R.string.file_explorer_message2);
                    }
                    Log.i("elapsed", String.format("meta %d", stopWatch.getTime()));
                    stopWatch.reset();
                    stopWatch.start();
                    GpsDirectory gpsDirectory = metadata.getFirstDirectoryOfType(GpsDirectory.class);
                    if (gpsDirectory != null && gpsDirectory.getGeoLocation() != null) {
                        item.longitude = gpsDirectory.getGeoLocation().getLongitude();
                        item.latitude = gpsDirectory.getGeoLocation().getLatitude();
                        List<Address> listAddress = CommonUtils.getFromLocation(BatchPopupActivity.this, item.latitude, item.longitude, 1, 0);
                        if (listAddress.size() > 0) {
                            item.info = CommonUtils.fullAddress(listAddress.get(0));
                        }
                        Log.i("elapsed", String.format("geo coding %d", stopWatch.getTime()));
                        ArrayList<PhotoMapItem> tempList = PhotoMapDbHelper.selectPhotoMapItemBy("imagePath", item.imagePath);
                        stopWatch.reset();
                        stopWatch.start();
                        if (tempList.size() > 0) {
                            mReduplicationCount++;
                        } else {
                            PhotoMapDbHelper.insertPhotoMapItem(item);
                            BitmapUtils.createScaledBitmap(targetFile.getAbsolutePath(), Constant.WORKING_DIRECTORY + fileName + ".thumb", 200);
                            mSuccessCount++;
                            Log.i("elapsed", String.format("create bitmap %d", stopWatch.getTime()));
                            stopWatch.stop();
                        }
                    } else {
                        mNoGPSInfoCount++;
                    }

                } catch (Exception e) {
                    mFailCount++;
                }
                progressHandler.sendMessage(message);
            }
        }
    }

    @Override
    public void onBackPressed() {
        DialogUtils.showAlertDialog(BatchPopupActivity.this, getString(R.string.batch_popup_message6));
    }

}

File: app/src/main/java/me/blog/korn123/easyphotomap/activities/BatchPopupActivity.kt
package me.blog.korn123.easyphotomap.activities

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Typeface
import android.os.Bundle
import android.os.Handler
import io.github.aafactory.commons.utils.BitmapUtils
import kotlinx.android.synthetic.main.activity_batch_popup.*
import me.blog.korn123.easyphotomap.R
import me.blog.korn123.easyphotomap.extensions.config
import me.blog.korn123.easyphotomap.extensions.showAlertDialog
import me.blog.korn123.easyphotomap.helper.COLUMN_IMAGE_PATH
import me.blog.korn123.easyphotomap.helper.PHOTO_MAP_THUMBNAIL_FIXED_WIDTH_HEIGHT
import me.blog.korn123.easyphotomap.helper.PhotoMapDbHelper
import me.blog.korn123.easyphotomap.helper.WORKING_DIRECTORY
import me.blog.korn123.easyphotomap.models.PhotoMapItem
import me.blog.korn123.easyphotomap.utils.EasyPhotoMapUtils
import org.apache.commons.io.FileUtils
import org.apache.commons.io.FilenameUtils
import java.io.File
import java.util.*

/**
 * Created by CHO HANJOONG on 2016-09-11.
 */
class BatchPopupActivity : SimpleActivity() {
    private var mEnableUpdate = true
    private var mTotalPhoto = 0
    private var mSuccessCount = 0
    private var mFailCount = 0
    private var mNoGPSInfoCount = 0
    private var mReduplicationCount = 0
    private var mProgressStatus = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_batch_popup)
        infoText.typeface = Typeface.DEFAULT
        infoText2.typeface = Typeface.DEFAULT
        infoText3.typeface = Typeface.DEFAULT
        infoText4.typeface = Typeface.DEFAULT
        infoText5.typeface = Typeface.DEFAULT
        infoText6.typeface = Typeface.DEFAULT

        val listImagePath = when (intent.getStringExtra("currentPath") == null) {
            true -> intent.getStringArrayListExtra("listImagePath")
            false -> {
                val targetPath = intent.getStringExtra("currentPath")
                determinePhotoFiles(File(targetPath))
                listFilePath
            }
        }

        mTotalPhoto = listImagePath.size
        when (mTotalPhoto > 0) {
            true -> {
                if (!File(WORKING_DIRECTORY).exists()) {
                    File(WORKING_DIRECTORY).mkdirs()
                }
                val thread = RegisterThread(this@BatchPopupActivity, listImagePath)
                progressBar.max = mTotalPhoto
                thread.start()
                stop.setOnClickListener({ _ -> mEnableUpdate = false })
                close.setOnClickListener({ _ -> finish() })
            }
            false -> {
                val builder = AlertDialog.Builder(this@BatchPopupActivity)
                val positiveListener = DialogInterface.OnClickListener { _, _ ->
                    finish()
                    return@OnClickListener
                }
                builder.setMessage(getString(R.string.file_explorer_message9))
                builder.setPositiveButton(getString(R.string.ok), positiveListener)
                val alertDialog = builder.create()
                alertDialog.show()
            }
        }
    }

    private val listFilePath: java.util.ArrayList<String> = arrayListOf()
    private fun determinePhotoFiles(directory: File) {
        directory.listFiles()?.map { file ->
            when (file.isFile) {
                true -> {
                    if (file.absoluteFile.extension.toLowerCase().matches("jpg|jpeg".toRegex())) {
                        listFilePath.add(file.absolutePath)
//                        println(file.absolutePath)
                    }
                }
                false -> determinePhotoFiles(file)
            }
        }
    }

    internal var progressHandler = Handler(Handler.Callback {
        infoText.text = "${++mProgressStatus} / $mTotalPhoto"
        infoText2.text = "${getString(R.string.batch_popup_message2)}: $mSuccessCount"
        infoText3.text = "${getString(R.string.batch_popup_message3)}: $mReduplicationCount"
        infoText4.text = "${getString(R.string.batch_popup_message4)}: $mNoGPSInfoCount"
        infoText5.text = "${getString(R.string.batch_popup_message5)}: $mFailCount"
        progressBar.progress = mProgressStatus
        true
    })

    internal inner class RegisterThread(var context: Context, private var listImagePath: ArrayList<String>) : Thread() {

        override fun run() {
            for (imagePath in listImagePath) {
                
                if (!mEnableUpdate) return
                val message = progressHandler.obtainMessage()
                try {
                    var fileName = FilenameUtils.getName(imagePath) + ".origin"
                    var targetFile: File?
                    if (config.enableCreateCopy) {
                        targetFile = File(WORKING_DIRECTORY + fileName)
                        if (!targetFile.exists()) {
                            FileUtils.copyFile(File(imagePath), targetFile)
                        }
                    } else {
                        targetFile = File(imagePath)
                        // remove .origin extension
                        fileName = FilenameUtils.getBaseName(fileName)
                    }

                    if (PhotoMapDbHelper.selectPhotoMapItemBy(COLUMN_IMAGE_PATH, targetFile.absolutePath).size > 0) {
                        mReduplicationCount++
                    } else {
                        val exifInfo = EasyPhotoMapUtils.parseExifDescription(targetFile.absolutePath)
                        val item = PhotoMapItem()
                        item.imagePath = targetFile.absolutePath
                        if (exifInfo.date != null) {
                            item.date = EasyPhotoMapUtils.dateTimePattern.format(exifInfo.date)
                        } else {
                            item.date = getString(R.string.file_explorer_message2)
                        }

                        exifInfo.geoLocation?.let {
                            item.longitude = it.longitude
                            item.latitude = it.latitude
                            val listAddress = EasyPhotoMapUtils.getFromLocation(this@BatchPopupActivity, item.latitude, item.longitude, 1, 0)
                            listAddress?.let {
                                if (it.isNotEmpty()) item.info = EasyPhotoMapUtils.fullAddress(it[0])
                            }
                            PhotoMapDbHelper.insertPhotoMapItem(item)
                            val srcPath = targetFile.absolutePath
                            Thread(Runnable {
                                BitmapUtils.saveBitmap(srcPath, WORKING_DIRECTORY + fileName + ".thumb", PHOTO_MAP_THUMBNAIL_FIXED_WIDTH_HEIGHT, exifInfo.tagOrientation)
                            }).start()
                            mSuccessCount++
                        } ?: mNoGPSInfoCount++
                        
                    }
                } catch (e: Exception) {
                    mFailCount++
                }
                progressHandler.sendMessage(message)
            }
        }
    }

    override fun onBackPressed() {
        showAlertDialog(getString(R.string.batch_popup_message6))
    }
}
